lib/lambda_ethereum_consensus/validator/validator.ex

```diff
@@ -408,35 +411,69 @@ defmodule LambdaEthereumConsensus.Validator do
   # Returns true if the validator should propose a block for the given slot
   defp should_propose?(%{duties: %{proposer: slots}}, slot), do: Enum.member?(slots, slot)
 
-  defp propose(%{root: head_root, validator: %{index: index, privkey: privkey}}, proposed_slot) do
-    head_state = BlockStates.get_state!(head_root)
-
-    # ExecutionClient.notify_forkchoice_updated()
-
-    execution_payload = Proposer.get_execution_payload()
-
-    block_request =
-      %BlockRequest{
-        slot: proposed_slot,
-        proposer_index: index,
-        graffiti_message: @default_graffiti_message,
-        eth1_data: fetch_eth1_data(proposed_slot, head_state),
-        execution_payload: execution_payload
-      }
-      |> Map.merge(Proposer.fetch_operations_for_block())
+  @spec build_execution_block(BeaconState.t(), Types.root()) ::
+          {:error, any()} | {:ok, Types.ExecutionPayload.t()}
+  def build_execution_block(state, head_root) do
+    head_block = Blocks.get_block!(head_root)
+    finalized_block = state.finalized_checkpoint.root |> Blocks.get_block!()
+
+    forkchoice_state = %{
+      finalized_block_hash: finalized_block.body.execution_payload.block_hash,
+      head_block_hash: head_block.body.execution_payload.block_hash,
+      # TODO calculate safe block
+      safe_block_hash: finalized_block.body.execution_payload.block_hash
+    }
 
-    # TODO: handle errors if there are any
-    {:ok, signed_block} = Proposer.construct_block(head_state, block_request, privkey)
+    payload_attributes = %{
+      timestamp: Misc.compute_timestamp_at_slot(state, state.slot),
+      prev_randao: Randao.get_randao_mix(state.randao_mixes, Accessors.get_current_epoch(state)),
+      suggested_fee_recipient: <<0::160>>,
+      # TODO: add withdrawals
+      withdrawals: [],
+      parent_beacon_block_root: head_root
+    }
 
-    {:ok, ssz_encoded} = Ssz.to_ssz(signed_block)
-    {:ok, encoded_msg} = :snappyer.compress(ssz_encoded)
-    fork_context = BeaconChain.get_fork_digest() |> Base.encode16(case: :lower)
+    with {:ok, payload_id} <-
+           ExecutionClient.notify_forkchoice_updated(
+             forkchoice_state,
+             payload_attributes
+           ),
+         # TODO: find a better way
+         :ok <- Process.sleep(3000),
```

Comment:
We should retry this a few times:

```elixir
Stream.repeatedly(fn -> ExecutionClient. get_payload(payload_id) end)
|> Stream.zip(Stream.timer(timeout_in_ms) |> Stream.cycle())
|> Stream.map(fn {v, 0} -> v end) 
|> Stream.take(number_of_retries)
|> Enum.find(&match?({:ok, _}, &1})
```

Reply:
But isn't that time variable? What happens if we call this without enough sleep?

Reply:
There will be fewer/no transactions but it will succeed

Reply:
When we make this code more sophisticated, we need to balance a time that should be long enough to let the execution pack as many transactions as possible (more fees for us) while giving enough time to propagate the block and have it included

Reply:
Added this comment in code.